Cisco NXOS Driver

hakasapl opened the pull request:
This PR implements a cisco NXOS driver for ansible-switches. The idea is that we should be able to write manifests the same way for both dell os9 and cisco NXOS and have it behave the same. This requirement is fulfilled with this driver, minus a couple small parity things that I will update on the os9 driver later.

The way it works is really different from the os9 driver because cisco has better functions for this. The general process is:
1. Ansible gets the existing running-config on the switch and a function in the python driver removes all lines that are managed by the manifest. This will include all vlan and interface lines, except if an interface is marked as managed or it doesn't exist in the manifest, the existing config for that interface will not be removed.
2. A different python function then takes the defined interfaces and vlans from the manifests and generates a cisco NXOS configuration in the correct format and appends it to the bottom of the original filtered running config. NXOS is smart so this sometimes results in duplicate interface sections, but NXOS concatenates these while applying.
3. This new config is then uploaded to the switch
4. NXOS has a config replace feature where it can replace the running config without causing downtime or requiring a reload. That command is run on the switch. It also has a commit-timeout feature, which the driver sets to 60 seconds. That means that if the driver doesn't commit the replaced config within 60 seconds of applying, it reverts (this is to prevent management lockouts from misconfigured manifests)
5. After the config is applied if it succeeded the driver commits the replaced config to avoid the rollback and saves the changes

@naved001 the core switches in the updated hosts file in this PR are not being used if you wanted to test against them. I suspect bugs will come up as we deploy stuff but I tried a bunch of stuff in the meantime and it seems to be working for me.

naved001 commented on filter_plugins/cisco_nxos.py:
```diff
@@ -0,0 +1,370 @@
+
+            # Check if object is configured
+            if object_name.count("/") > 1:
+                # this is a breakout
```
is this assuming that `Ethernet1/1` means the parent/whole interface but `Ethernet1/1/3` means a breakout interface? Is this mentioned in any nx-os documentation that this is a convention that's followed?

I vaguely remember cisco switches with interface names like `gi 1/0/1` and those weren't breakout ports. I think it used to donate slot or modules.

hakasapl replied:
Yeah we'll very likely never have devices with slots like that but it's a good point. I guess the best way then is to see if 1/29 already exists, and if 1/29/1 is found when 1/29 was already found then its a breakout

naved001 commented on filter_plugins/cisco_nxos.py:
```diff
@@ -0,0 +1,370 @@
+                object_name = object_name[:object_name.rfind("/")]
+
+            if object_name in interfaces:
+                if "managed" in interfaces[object_name] and interfaces[object_name]["managed"]:
```
you can almost always simplify this to:
```python
if interfaces[object_name].get("managed"):
```
`.get()` on a dictionary will return `None` if the key doesn't exist which is Falsy in python

hakasapl replied:
Did not know that, thanks!

naved001 commented on filter_plugins/cisco_nxos.py:
```diff
@@ -0,0 +1,370 @@
+                    continue
+            else:
+                new_config.append(line)
+                delete_section = False
```
I am having a hard time following the purpose of delete_section and how we are deciding to set it or unset it.

hakasapl replied:
I will add some stuff in the dockstring and some comments

naved001 commented on filter_plugins/cisco_nxos.py:
```diff
@@ -0,0 +1,370 @@
+
+        config.append("")  # Newline
+        config.append(f"vlan {str(id)}")
+        config.append(f"  name {fields["name"]}")
```
syntax error here because you are using the same quotes.
```
    config.append(f" name {fields["name"]}")
                                   ^
SyntaxError: f-string: unmatched '['
```

I would create a multi-line string instead of calling config.append mutliple lines.

hakasapl replied:
I think it's okay to use double quotes like this because python actually parses the {} separately. It works fine like that when I run it.

I will make it multiline

hakasapl replied:
Well, actually if you got a syntax error I'm not sure, maybe it's just the way ansible runs filter plugins that works better? I'll just make them single quotes

naved001 replied:
I am kind of surprised that this works. Anyway, it fails for me.
```
➜  ~ cat test.py
fields = {}
config = {}
id = "23"
fields["name"] = "foo"

config.append("")  # Newline
config.append(f"vlan {str(id)}")
config.append(f"  name {fields["name"]}")
config.append(f" name {fields["name"]}")

print(lines)

➜  ~ python test.py
  File "/Users/naved/test.py", line 8
    config.append(f"  name {fields["name"]}")
                                    ^
SyntaxError: f-string: unmatched '['
```

larsks replied:
That is a python version issue. The syntax is correct with current (3.12) python. Not sure exactly when it became valid.

naved001 reviewed (changes requested):
Just some nitpicking and a bunch of questions.  I think it would be great to add some unit tests for the filter plugins. I will continue to review the rest of it later.

hakasapl commented on filter_plugins/cisco_nxos.py:
```diff
@@ -0,0 +1,370 @@
+import re
+
+def parseRanges(config):
```
Yes, will add that

hakasapl commented on filter_plugins/cisco_nxos.py:
```diff
@@ -0,0 +1,370 @@
+            for vlan_part in vlan_range_str.split(","):
+                if "-" in vlan_part:
+                    vlan_range = vlan_part.split("-")
+                    vlan_list += [str(i) for i in range(int(vlan_range[0]), int(vlan_range[-1]) + 1)]
```
I figured it was more readable as "first" and "last", no strong preference either way

naved001 replied:
ah, I see. I'd probably do the split like this if we want to make it more readable
```python
>>> first, last = "24-28".split("-")
>>> first
'24'
>>> last
'28'
```

larsks replied:
Just for kicks, you can also solve this using purely functional programming:

```python
def split_vlan_range(vlan_range_str: list[str]):
    return [
        vlan
        for vlan_span in [
            range(int(y[0]), int(y[1]) + 1)
            for y in [
                x.split("-") if "-" in x else (x, x) for x in vlan_range_str.split(",")
            ]
        ]
        for vlan in vlan_span
    ]
```

In use:

```
>>> split_vlan_range('1,2,3')
[1, 2, 3]
>>> split_vlan_range('1,2-5,6')
[1, 2, 3, 4, 5, 6]
>>> split_vlan_range('1')
[1]
```

hakasapl commented on filter_plugins/cisco_nxos.py:
```diff
@@ -0,0 +1,370 @@
+
+    return vlan_list
+
+def removeOldLines(config, interfaces, vlans):
```
Yes, I forgot about docstrings for everything so I will do that for all the methods

larsks commented on filter_plugins/cisco_nxos.py:
```diff
@@ -0,0 +1,443 @@
+    boot nxos bootflash:/nxos64-cs.10.3.2.F.bin
+    no system default switchport shutdown
+
+    Output of this mehtod will be:
```
Typo, `s/mehtod/method/`

larsks commented on filter_plugins/cisco_nxos.py:
```diff
@@ -0,0 +1,443 @@
+    Parses vlan ranges in config files. For example, vlan 5-8,10,15-17 will return [5,6,7,8,10,15,16,17]
+
+    :param config: The config string block to parse the VLAN block for
+    :type config: str
```
Type annotations are a great way to document the type of parameters:

```
def getVLANList(config: str):
    ...
```

(And allow things to be type-checked in your development environemnt.)

larsks reviewed (commented):
Looks good. There is a lot of logic in here; I would love to see some unit tests to go with it. Maybe something like [this](https://github.com/larsks/ansible-switches/compare/cisco...larsks:ansible-switches:example/tests?expand=1)?

I've noted a harmless typo, and I've made a couple of style comments that you are free to ignore.

larsks commented on filter_plugins/cisco_nxos.py:
```diff
@@ -0,0 +1,443 @@
+import re
+
+def getVLANList(config):
```
I don't know cisco config syntax. Should `getVLANList("vlan 1")` return `["1"]` or `[]`?

larsks replied:
@hakasapl wanted to draw your attention to this; is the current code wrong, or will we never see a `vlan` line without commas?

hakasapl replied:
Even if there isn't a comma shouldn't split() then just output a list of size 1 and work anyway? In the latest commit I moved the default vlan stuff around a bit to make more sense. That's added in the main filter plugin method though

larsks commented on filter_plugins/cisco_nxos.py:
```diff
@@ -0,0 +1,443 @@
+        if "fanout" in fields:
+            fanout_fields = fields["fanout"]
+            nxos_mode_str = ""
+            if fanout_fields["mode"] == "quad":
```
This appears to want field `mode`, but all the `fanout` sections in `host_vars` have `type` instead:

```
interfaces:
  hundredGigE 1/1:
    fanout:
      type: "quad"
      speed: "25G"
```

larsks commented on filter_plugins/cisco_nxos.py:
```diff
@@ -0,0 +1,443 @@
+        if dict_key not in config_dict:
+            config_dict[dict_key] = []
+
+        is_portchannel = intf.startswith("port-channel")
```
Shouldn't this be spelled `Port-channel`? In the config files, I see e.g.:

```
Port-channel 10:
  ...
Port-channel 11:
  ...
```

And I don't see anywhere we convert that to lower-case.

larsks replied:
Got it. I guess we don't have any cisco configs in `host_vars` at the moment.

naved001 commented:
@hakasapl a couple of high level questions:

1. Would it make sense to get the switch data in json format? `show running-config | json` returns a nice json which makes more sense to parse programmatically.
2. The switches support NX-API, is there a reason why the current method is preferred over the API?

hakasapl commented:
> @hakasapl a couple of high level questions:
> 
>     1. Would it make sense to get the switch data in json format? `show running-config | json` returns a nice json which makes more sense to parse programmatically.
> 
>     2. The switches support NX-API, is there a reason why the current method is preferred over the API?

A few things:
- The JSON output takes ages to run, like 60+ seconds for me as opposed to 2 seconds.
- The JSON output is a bit weird, a bunch of extra keys that I thought may not be able to be kept general across switches, but I can't say if I'm 100% right on that until I see what it looks like on a different switch
- The config replace command expects the normal format for configuration as far as I can tell, which means I'm going to have to switch to the regular readable format at some step anyway, so it didn't feel worth it to start with JSON
- We can switch it NX-API pretty much whenever we want, the ansible module supports both, initially I just did it with SSH because that's what os9 uses, but I think just changing the connection type is all that is required for that.

naved001 commented:
> We can switch it NX-API pretty much whenever we want, the ansible module supports both, initially I just did it with SSH because that's what os9 uses, but I think just changing the connection type is all that is required for that.

I wasn't suggesting to use the API to gather the config file and upload it back, but rather to make the actual change we want to make. But now that I think about it, it makes sense to upload the whole file since we read the whole state of things from the manifests (instead of the ESI model where a user is making specific changes at a time). So, I guess it doesn't matter in this case.

hakasapl commented:
> > We can switch it NX-API pretty much whenever we want, the ansible module supports both, initially I just did it with SSH because that's what os9 uses, but I think just changing the connection type is all that is required for that.
> 
> I wasn't suggesting to use the API to gather the config file and upload it back, but rather to make the actual change we want to make. But now that I think about it, it makes sense to upload the whole file since we read the whole state of things from the manifests (instead of the ESI model where a user is making specific changes at a time). So, I guess it doesn't matter in this case.

Right, config replace needs the whole config, modified with manifests, so we aren't making individual changes (that would take ages anyway)

hakasapl commented:
@larsks @naved001 I think I've addressed everything in your comments except unit tests. I would like to merge this and make a separate issue for the unit tests since this is already probably an 11 in my sprint. Thank you @larsks for the example tests

larsks commented:
I am really uncomfortable merging code without tests. I think that as a rule tests should be a fundamental part of our development process. If you'd like, we can add my example tests to this PR and then improve upon those in a subsequent change.

hakasapl commented:
> I am really uncomfortable merging code without tests. I think that as a rule tests should be a fundamental part of our development process. If you'd like, we can add my example tests to this PR and then improve upon those in a subsequent change.

Okay, I merged your example tests with my branch, and ran it on my local system and everything seems to be working correctly.

Unfortunately writing tests is always an afterthought for this project specifically because I almost always don't know what is "correct" until I see how it behaves on the switch. I will make a separate issue for fleshing out the unit tests for not only this driver but also dell os9, which does not have any tests currently

hakasapl commented:
Issue created here: https://github.com/CCI-MOC/ops-issues/issues/1377

Thanks again for your examples @larsks !

larsks commented:
@hakasapl I have fixed two of the tests and fixed the commit message on my "example tests" commit. You'll want to sync up with the most recent changes.

hakasapl commented:
> @hakasapl I have fixed two of the tests and fixed the commit message on my "example tests" commit. You'll want to sync up with the most recent changes.

Okay, I did that and also fixed the bug you noted above, so `vlan 1` gets parsed correctly. I also changes the test that expects that case to work

larsks reviewed (approved):
Thanks for fixing that up!

naved001 commented on filter_plugins/cisco_nxos.py:
```diff
@@ -0,0 +1,444 @@
+            # Find the object name
+            object_name = line.removeprefix("interface ")
+            if not object_name.startswith("Ethernet") and not object_name.startswith("port-channel"):
+                # This allows things like management interfaces to pass through
```
wouldn't this also allow vlan interfaces to passthrough? I assumed that we want to manage the vlan interfaces with these manifests?

hakasapl replied:
as far as I can tell nxos doesn't have vlan interfaces, everything is configured on the vlan itself, so it would show up as `vlan XYZ` in the config with `interface` in the beginning

naved001 replied:
It does support vlan interfaces. Our older cisco switches support it so that's why I was expecting it. But the same commands did not run on the new cisco switches, so I looked up [the documentation](https://www.cisco.com/c/en/us/td/docs/switches/datacenter/nexus9000/sw/7-x/interfaces/configuration/guide/b_Cisco_Nexus_9000_Series_NX-OS_Interfaces_Configuration_Guide_7x/b_Cisco_Nexus_9000_Series_NX-OS_Interfaces_Configuration_Guide_7x_chapter_0101.html#task_F0C90E46B5D84349BB62D68BC466B7CB) and then I learnt that we have to enable vlan interfaces first, and then it works as expected.

I enabled it with `feature interface-vlan` on the NERC-CORE-1 and then you can create vlan interfaces. Here's an example I created:

```
NERC-CORE-1(config)# feature inter
interface-vlan   intersight
NERC-CORE-1(config)# feature interface-vlan ?
  <CR>

NERC-CORE-1(config)# feature interface-vlan
NERC-CORE-1(config)# inter
interface    intersight
NERC-CORE-1(config)# interface vlan 201?
  ,         Multi range separator
  -         Range separator
  <1-4094>  Vlan interface number

NERC-CORE-1(config)# interface vlan 201
```

as soon as I enabled it, it also created a default interface for vlan 1.

```
NERC-CORE-1# show running-config |grep "interface Vlan"
interface Vlan1
interface Vlan201
```

feel free to delete the vlan interfaces or disable it for now.

naved001 commented on filter_plugins/cisco_nxos.py:
```diff
@@ -0,0 +1,444 @@
+        # "untagged" field
+        if "untagged" in fields:
+            untagged_str = ""
+            if fields["portmode"] == "hybrid":
```
in [line 307](https://github.com/CCI-MOC/ansible-switches/pull/59/files#diff-dd31f0cdd8a2ec6ba878bc26867abc7a3e3e1f6ba03d919eb9c7ab147990d1ecR307) we allow portmode to be either `trunk` or `hybrid` but here we expect it be only `hybrid`. So, if the manifests said "trunk", my switchport will be put in trunk mode but then this will not set the trunk native vlan and instead it'll set access vlan which would do nothing.

naved001 replied:
sorry I still don't understand what you mean here. My question is why can't we do this:

```python
if "untagged" in fields:
    untagged_str = ""
    if fields["portmode"] in ["hybrid", "trunk"]:
        untagged_str = f"switchport trunk native vlan {fields['untagged']}"
    elif fields["portmode"] == "access":
        untagged_str = f"switchport access vlan {fields['untagged']}"
```


so we effectively treat hybrid as trunk

hakasapl replied:
Because I don't want portmode "trunk" and "untagged" to be a valid schema supplied together. It's not a valid schema for OS9 so I wanted to go with the minimum there. In other words the manifests are validated agnostic of the switch OS, so the validation needs to work for every switch OS

naved001 replied:
ah, so OS9 expects only tagged in trunk mode. I guess we will deal with this later then and it's fine for now.

naved001 commented on filter_plugins/cisco_nxos.py:
```diff
@@ -0,0 +1,444 @@
+            if fields["portmode"] == "hybrid":
+                untagged_str = f"switchport trunk native vlan {fields['untagged']}"
+            else:
+                untagged_str = f"switchport access vlan {fields['untagged']}"
```
I think we should explicitly check that portmode is access before we set this

hakasapl replied:
I will make that change

naved001 commented on roles/common/tasks/cisco_nxos.yaml:
```diff
@@ -0,0 +1,78 @@
+    state: enabled
+  notify: Save Config
+
+- name: Enable lacp feature
```
@hakasapl you probably want to enable the vlan interface feature as part of this playbook then.

hakasapl replied:
Yeah definitely
